ESP/lib/src/local/network/stream/stream_server.cpp

```diff
@@ -0,0 +1,252 @@
+//!                             Global Variables
+// ***********************************************************************************
+
+constexpr static const char* STREAM_CONTENT_TYPE =
+    "multipart/x-mixed-replace;boundary=" PART_BOUNDARY;
+constexpr static const char* STREAM_BOUNDARY = "\r\n--" PART_BOUNDARY "\r\n";
+/* constexpr static const char* STREAM_PART =
+    "Content-Type: image/jpeg\r\nContent-Length: %u\r\nX-Timestamp: "
+    "%d.%06d\r\n\r\n";
+ */
+constexpr static const char* STREAM_PART =
+    "Content-Type: %s\r\nContent-Length: %u\r\n\r\n";
+static const char* JPG_CONTENT_TYPE = "image/jpeg";
+
+// ***********************************************************************************
+//!                             AsyncBufferResponse
+// ***********************************************************************************
+
+AsyncBufferResponse::AsyncBufferResponse(uint8_t* buf,
+                                         size_t len,
+                                         const char* contentType)
+    : _buf(buf), _len(len) {
+  _callback = nullptr;
+  _code = 200;
+  _contentLength = _len;
+  _contentType = contentType;
+  _index = 0;
+}
+
+AsyncBufferResponse::~AsyncBufferResponse() {
+  if (_buf != nullptr) {
+    free(_buf);
+  }
+}
+
+bool AsyncBufferResponse::_sourceValid() const {
+  return _buf != nullptr;
+}
+
+size_t AsyncBufferResponse::_fillBuffer(uint8_t* buf, size_t maxLen) {
+  size_t ret = _content(buf, maxLen, _index);
+  if (ret != RESPONSE_TRY_AGAIN) {
+    _index += ret;
+  }
+  return ret;
+}
+
+size_t AsyncBufferResponse::_content(uint8_t* buffer,
+                                     size_t maxLen,
+                                     size_t index) {
+  memcpy(buffer, _buf + index, maxLen);
+  if ((index + maxLen) == _len) {
+    free(_buf);
+    _buf = nullptr;
+  }
+  return maxLen;
+}
+
+// ***********************************************************************************
+//!                             AsyncFrameResponse
+// ***********************************************************************************
+
+AsyncFrameResponse::AsyncFrameResponse(camera_fb_t* frame,
+                                       const char* contentType)
+    : fb(frame), _index(0) {
+  _callback = nullptr;
+  _code = 200;
+  _contentLength = frame->len;
+  _contentType = contentType;
+}
+
+AsyncFrameResponse::~AsyncFrameResponse() {
+  if (fb != nullptr) {
+    esp_camera_fb_return(fb);
+  }
+}
+
+bool AsyncFrameResponse::_sourceValid() const {
+  return fb != nullptr;
+}
+
+size_t AsyncFrameResponse::_content(uint8_t* buffer,
+                                    size_t maxLen,
+                                    size_t index) {
+  memcpy(buffer, fb->buf + index, maxLen);
+  if ((index + maxLen) == fb->len) {
+    esp_camera_fb_return(fb);
+    fb = nullptr;
+  }
+  return maxLen;
+}
+
+size_t AsyncFrameResponse::_fillBuffer(uint8_t* buf, size_t maxLen) {
```

Comment:
I'm wondering if we need those, the AsyncFrameResponse and AsyncBufferResponse. I don't see them being used for streaming or API responses, are they needed internally somewhere or?

Reply:
I implemented these so that we can take images, instead of video. To use as previews, it would be more efficient. They are not currently implemented, just there holding space in case we need them.

Reply:
gotcha, I actually thought we we'd just take a still from the stream but that's also fine :D